### DISCUSSION TITLE
How to convert Any image to gif file (no-anime) ?

### DISCUSSION THREAD

**Discussion by IOrlandoni** [libvips/php-vips#184] [Q&A]
How can I convert any image file to a non-animated gif, like PNG to JPG? I always get an error.

**Comment by jcupitt** [comment]
Hi @web-apply,

It should just be:

```php
$im = Vips\Image::newFromFile("x.png", ['access' => Vips\Access::SEQUENTIAL]);
$im->writeToFile("y.gif");
```

You'll need to give a lot more information: platform, image samples, libvips version, php-vips version, install method, error message, etc. etc.

**Comment by IOrlandoni** [comment]
@jcupitt  I couldn't set FFI in PHP. Because I'm using Litespeed PHP (LSAPI). So I see my work with legacy code. Problem only writing gif file.

 $x = vips_image_new_from_file("is/a.jpg")["out"];
ok - vips_image_write_to_file($x, "out/e.tif",["Q"=>90]);
ok - vips_image_write_to_file($x, "out/e.JPG",["Q"=>90]);
ok - vips_image_write_to_file($x, "out/e.heic",["Q"=>90]);
ok - vips_image_write_to_file($x, "out/e.png",["Q"=>90]);
ok - vips_image_write_to_file($x, "out/e.webp",["Q"=>90]);
ok - vips_image_write_to_file($x, "out/e.jfif",["Q"=>90]);

**no working - vips_image_write_to_file($x, "out/e.gif",["Q"=>90]);
  $x = vips_call("copy", $x)["out"];
Not working: vips_image_write_to_file($x, "out/e.gif",["Q"=>50]);**   



 VIP version: 8.13.2 

Php Ext functions 
(
--
  | [0] => vips_image_new_from_file
  | [1] => vips_image_new_from_buffer
  | [2] => vips_image_new_from_array
  | [3] => vips_image_write_to_file
  | [4] => vips_image_write_to_buffer
  | [5] => vips_image_copy_memory
  | [6] => vips_image_new_from_memory
  | [7] => vips_image_write_to_memory
  | [8] => vips_image_write_to_array
  | [9] => vips_foreign_find_load
  | [10] => vips_foreign_find_load_buffer
  | [11] => vips_interpolate_new
  | [12] => vips_call
  | [13] => vips_image_get
  | [14] => vips_image_get_typeof
  | [15] => vips_image_set
  | [16] => vips_type_from_name
  | [17] => vips_image_set_type
  | [18] => vips_image_remove
  | [19] => vips_error_buffer
  | [20] => vips_cache_set_max
  | [21] => vips_cache_set_max_mem
  | [22] => vips_cache_set_max_files
  | [23] => vips_concurrency_set
  | [24] => vips_cache_get_max
  | [25] => vips_cache_get_max_mem
  | [26] => vips_cache_get_max_files
  | [27] => vips_cache_get_size
  | [28] => vips_concurrency_get
  | [29] => vips_version
  | )

**Comment by jcupitt** [comment]
You can use the 1.x branch of php-vips and get a nice API. I wouldn't use php-vips-ext directly.

What libvips version is this? If it's older than 8.11, you need to update to get GIF write.

**Reply by IOrlandoni** [reply]
@jcupitt VIPS version: 8.13.2

**Reply by jcupitt** [reply]
Have you tried removing the `Q` parameter? You don't need it for GIF.

**Reply by IOrlandoni** [reply]
yes,

```
   $x = vips_image_new_from_file("is/a.jpg")["out"];
         $x = vips_call("invert", $x)["out"];
	echo vips_image_write_to_file($x, "is/vava.gif"); 
```
	
	not saving gif file.
	
	but  on shell command - vips copy a.jpg  a.gif      working.  and on shell or php  vips version same. 8.13.2
	
	can i use  vips_call("copy") etc.   command with php ?

**Reply by IOrlandoni** [reply]
Could the problem be with the libvips42 package version?

https://packages.debian.org/bullseye/libvips42   

Package: libvips42 (8.10.5-2)

**Reply by jcupitt** [reply]
Ah, OK. 

Yes, php is probably picking up 8.10, not 8.13.

You can call `vips_version()` to get the real library version number.

**Reply by IOrlandoni** [reply]
@jcupitt thanks 
I found the problem. The problem was in the cgif library. I have not uploaded before. I have now installed and reinstalled libvips latest version.
Installed  via git clone.  8.15 is stable ?

**Reply by jcupitt** [reply]
8.14.1 is the current stable release:

https://github.com/libvips/libvips/releases

**Comment by dppham1** [comment]
@jcupitt I have the following code for converting HEIC -> JPG:

```
        // Write image to temp file
        $raw_photo = $request->getBody();
        $temp = tmpfile();
        fwrite($temp, $raw_photo);

        // Get the path
        $temp_meta = stream_get_meta_data($temp);
        $temp_path = $temp_meta['uri'];
    
        // Create image
        $image = Image::newFromFile($temp_path);
         
          // Convert to JPG
         $image->jpegsave($temp_path . '.jpg');        
        // $image->writeToFile($temp_path . '.jpg');      <-- I also tried this, but didn't work

```


The generated file does have a .jpg extension:

<img width="255" alt="Screenshot 2024-06-28 at 2 57 02 PM" src="https://github.com/libvips/php-vips/assets/15275375/d4b4bf49-18e6-4013-9a8a-9001f50a8ccf">

However, certain applications like Slack or even using https://www.checkfiletype.com shows the image as still being HEIF:
<img width="602" alt="Screenshot 2024-06-28 at 2 58 41 PM" src="https://github.com/libvips/php-vips/assets/15275375/b2321923-4825-434f-a8f6-608898980313">

Is there anything else I am missing?

**Reply by jcupitt** [reply]
Hi, it should just be:

```php
$im = Vips\Image::newFromFile("something.heic", ['access' => Vips\Access::SEQUENTIAL]);
$im->writeToFile("somethingelse.jpg");
```

You probably have a bug somewhere else in your code, or you copied the wrong file. 

You'd need to share a complete example that would let someone else reproduce the problem, and include details like platform, libvips version, any error message, etc. etc.

**Reply by jcupitt** [reply]
Also, you should open a new discussion for a new question, and you should not post the same question in many places, eg.:

https://stackoverflow.com/questions/78684618/php-libvips-convert-heic-to-jpg 

Asking the same question several times just makes unnecessary work for the volunteers who help support these projects.
